File: code/design-patterns/behavioral/mediator/python/mediator.py
from abc import ABC, abstractmethod
from typing import List


class Mediator(ABC):
    @abstractmethod
    def can_arrive(self, train: "Train") -> bool:
        pass

    @abstractmethod
    def notify_about_departure(self):
        pass


class StationManager(Mediator):
    def __init__(self):
        self.train_queue: List[Train] = []

    def can_arrive(self, train: "Train") -> bool:
        if not self.train_queue:
            self.train_queue.append(train)
            return True
        else:
            print(f"{train.__class__.__name__} is waiting for the platform.")
            return False

    def notify_about_departure(self):
        if self.train_queue:
            departed_train = self.train_queue.pop(0)
            print(f"{departed_train.__class__.__name__} has departed.")
        if self.train_queue:
            next_train = self.train_queue[0]
            print(f"{next_train.__class__.__name__} can now arrive.")
            next_train.arrive()


class Train(ABC):
    def __init__(self, mediator: Mediator):
        self.mediator = mediator

    @abstractmethod
    def arrive(self):
        pass

    @abstractmethod
    def depart(self):
        pass


# Client code is simple, just ask the mediator
class PassengerTrain(Train):
    def arrive(self):
        if self.mediator.can_arrive(self):
            print("Passenger train arrived.")
        else:
            print("Passenger train waiting for arrival.")

    def depart(self):
        print("Passenger train leaving...")
        self.mediator.notify_about_departure()


class FreightTrain(Train):
    def arrive(self):
        if self.mediator.can_arrive(self):
            print("Freight train arrived.")
        else:
            print("Freight train waiting for arrival.")

    def depart(self):
        print("Freight train leaving...")
        self.mediator.notify_about_departure()

File: code/design-patterns/behavioral/mediator/go/mediator.go
package main

import "fmt"

// Mediator interface
type Mediator interface {
	CanArrive(train Train) bool
	NotifyAboutDeparture()
}

// Concrete Mediator
type StationManager struct {
	trainQueue []Train
}

func NewStationManager() *StationManager {
	return &StationManager{}
}

func (s *StationManager) CanArrive(train Train) bool {
	if len(s.trainQueue) == 0 {
		s.trainQueue = append(s.trainQueue, train)
		return true
	} else {
		fmt.Printf("%T is waiting for the platform.\n", train)
		return false
	}
}

func (s *StationManager) NotifyAboutDeparture() {
	if len(s.trainQueue) > 0 {
		departedTrain := s.trainQueue[0]
		s.trainQueue = s.trainQueue[1:]
		fmt.Printf("%T has departed.\n", departedTrain)
	}
	if len(s.trainQueue) > 0 {
		nextTrain := s.trainQueue[0]
		fmt.Printf("%T can now arrive.\n", nextTrain)
		nextTrain.Arrive()
	}
}

// Train interface
type Train interface {
	Arrive()
	Depart()
}

// PassengerTrain struct
type PassengerTrain struct {
	mediator Mediator
}

func NewPassengerTrain(mediator Mediator) *PassengerTrain {
	return &PassengerTrain{mediator: mediator}
}

func (p *PassengerTrain) Arrive() {
	if p.mediator.CanArrive(p) {
		fmt.Println("Passenger train arrived.")
	} else {
		fmt.Println("Passenger train waiting for arrival.")
	}
}

func (p *PassengerTrain) Depart() {
	fmt.Println("Passenger train leaving...")
	p.mediator.NotifyAboutDeparture()
}

// FreightTrain struct
type FreightTrain struct {
	mediator Mediator
}

func NewFreightTrain(mediator Mediator) *FreightTrain {
	return &FreightTrain{mediator: mediator}
}

func (f *FreightTrain) Arrive() {
	if f.mediator.CanArrive(f) {
		fmt.Println("Freight train arrived.")
	} else {
		fmt.Println("Freight train waiting for arrival.")
	}
}

func (f *FreightTrain) Depart() {
	fmt.Println("Freight train leaving...")
	f.mediator.NotifyAboutDeparture()
}
